File: 81302_거리두기확인하기.py
from collections import deque
from pickle import TRUE
from tkinter.tix import Tree
MAXR = 5
MAXC = 5

def getManDistance(r1,c1, r2,c2):
    return abs(r1-r2) + abs(c1-c2)

def isOutOfRange(r,c):
    return (r < 0 or r >= MAXR or c < 0 or c >= MAXC)

def hasClosePeople(place, row, col):
    dr = [-1,1,0,0]
    dc = [0,0,-1,1]
    searchQ = deque()
    searchQ.append([row, col])
    while len(searchQ) > 0:
        currR, currC = searchQ.popleft()
        dis = getManDistance(row,col, currR, currC)

        if dis >= 2:
            continue
        searchPosList = [(currR+r,currC+c) for r,c in zip(dr,dc) if not isOutOfRange(currR+r, currC+c)]
        for nextR, nextC in searchPosList:
            if(nextR == row and nextC == col): continue
            if place[nextR][nextC] == 'P':
                return True
            
            if place[nextR][nextC] == 'O' and dis < getManDistance(row,col, nextR, nextC):
                searchQ.append([nextR, nextC])
            
    return False

def getSolByRoom(place):
    for row, rows in enumerate(place):
        for col, string in enumerate(rows):
            if(string == 'P'):
                if hasClosePeople(place, row, col):
                    return 0
    return 1

def solution(places):

    answer = [getSolByRoom(place) for place in places]
    print(answer)
    return answer


places = [["POOOP", "OXXOX", "OPXPX", "OOXOX", "POXXP"], ["POOPX", "OXPXP", "PXXXO", "OXXXO", "OOOPP"], ["PXOPX", "OXOXP", "OXPOX", "OXXOP", "PXPOX"], ["OOOXX", "XOOOX", "OOOXX", "OXOOX", "OOOOO"], ["PXPXP", "XPXPX", "PXPXP", "XPXPX", "PXPXP"]]
solution(places)

File: Programmers/81302_거리두기확인하기.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;
const int MAXR = 5;
const int MAXC = 5;
class Point
{
public:
    int r;
    int c;
    int dis;
    Point(int r, int c, int dis) : r(r), c(c), dis(dis)
    {

    }
};


int visit[MAXR][MAXC];

void initVisit()
{
    for (int r = 0; r < MAXR; r++)
        for (int c = 0; c < MAXC; c++)
            visit[r][c] = 0;
}

bool checkClosePerson(vector<string>& room, int r, int c)
{
    initVisit();
    const int dr[] = { -1,1,0,0 };
    const int dc[] = { 0,0,-1,1 };
    queue<Point> q;
    q.push(Point(r, c, 0));

    while (!q.empty())
    {
        Point temp = q.front();
        q.pop();

        visit[temp.r][temp.c] = 1;

        for (int i = 0; i < 4; i++)
        {
            int newr = temp.r + dr[i];
            int newc = temp.c + dc[i];
            int newDis = temp.dis + 1;

            if (newr < 0 || newr >= MAXR || newc < 0 || newc >= MAXC)
                continue;

            if (visit[newr][newc] == 1) continue;
            if (newDis > 2) continue;
            if (room[newr][newc] == 'P')
                return true;

            if (room[newr][newc] == 'O' )
                q.push(Point(newr, newc, newDis));


        }

    }
    return false;
}


int getSolutionByRoom(vector<string>& room)
{
    for (int r = 0; r < MAXR; r++)
    {
        for (int c = 0; c < MAXC; c++)
        {
            if (room[r][c] == 'P' && checkClosePerson(room, r, c))
            {
                return 0;
            }
        }
    }
    return 1;
}

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    for (auto room : places)
    {
        answer.push_back(getSolutionByRoom(room));
    }
    return answer;
}


vector<vector<string>> input = { {"POOOP", "OXXOX", "OPXPX", "OOXOX", "POXXP"}, {"POOPX", "OXPXP", "PXXXO", "OXXXO", "OOOPP"}, {"PXOPX", "OXOXP", "OXPOX", "OXXOP", "PXPOX"}, {"OOOXX", "XOOOX", "OOOXX", "OXOOX", "OOOOO"}, {"PXPXP", "XPXPX", "PXPXP", "XPXPX", "PXPXP"} };


int main()
{
    solution(input);
    return 0;
}